File: core/src/main/kotlin/com/github/servb/pph/pheroes/common/common/NATION_TYPE.java
package com.github.servb.pph.pheroes.common.common;

import com.github.servb.pph.util.helpertype.EnumC;

/**
 *
 * @author SerVB
 */
public enum NATION_TYPE implements EnumC {

    NATION_NEUTRAL(0),
    NATION_HIGHMEN,
    NATION_BARBARIANS,
    NATION_WIZARDS,
    NATION_BEASTMEN,
    NATION_ELVES,
    NATION_UNDEADS,
    NATION_COUNT;

    //<editor-fold defaultstate="collapsed" desc="C-like enum (for indexing and count)">
    /**
     * The value of the element.
     */
    private int value;

    /**
     * Constructs a new element with the next value.
     */
    private NATION_TYPE() {
        this(NextValueHolder.nextValue); // Call the constructor with the next value
    }

    /**
     * Constructs a new element with the specified value.
     *
     * @param value The specified value.
     */
    private NATION_TYPE(final int value) {
        this.value = value;                     // Set the specified value to this value
        NextValueHolder.nextValue = value + 1;  // Increment the next value for a next element
    }

    /**
     * Returns the value of the element.
     *
     * @return The value of the element.
     */
    @Override
    public final int getValue() {
        return value;
    }

    /**
     * Object that holds the next value.
     */
    private final static class NextValueHolder {

        /**
         * The next value.
         */
        private static int nextValue = 0;

    }
    //</editor-fold>

}

File: core/src/main/kotlin/com/github/servb/pph/pheroes/common/common/NationType.kt
package com.github.servb.pph.pheroes.common.common

import com.github.servb.pph.util.helpertype.CountValueEnum
import com.github.servb.pph.util.helpertype.UniqueValueEnum

enum class NationType(
        override val v: Int,
        val mask: Int = -42,
        val ideology: IdeologyType? = null
) : UniqueValueEnum, CountValueEnum {
    NEUTRAL(0, ideology = IdeologyType.NEUTRAL),
    HIGHMEN(1, 0x1, IdeologyType.GOOD),
    BARBARIANS(2, 0x2, IdeologyType.EVIL),
    WIZARDS(3, 0x4, IdeologyType.GOOD),
    BEASTMEN(4, 0x8, IdeologyType.EVIL),
    ELVES(5, 0x10, IdeologyType.GOOD),
    UNDEADS(6, 0x20, IdeologyType.EVIL),
    COUNT(7);
}
